iOS Learning Group (#596)
artsy/artsy.github.io: _posts/2019-12-03-ios-learning-group.markdown

experiences to define and deliver a learning plan.

I started by booking five sessions, spread out by a week. I picked a time that was a good fit for as many engineers
as possible, and I scheduled them a month ahead to give people a chance to move their schedules around. Scheduling
them up front was important, to give learners a sense of what to expect; previous learning groups had suffered from
inconsistent schedules, which led to intermittent attendance. I also made sure to go to each Tech Lead and ask them
to make sure at least one engineer from their team attended.

Next was the actual curriculum. In the spirit of "proudly discovered elsewhere", we actually looked at using
existing learning resources that someone else had already developed. However, since
[our iOS stack is a bit unique](https://artsy.github.io/series/react-native-at-artsy/) and I already have
experience developing learning curricula, we decided to make our own. With the help of our Peer Learning Working
Group, we set learning objectives for the five sessions. Here was the original plan:

- **Week 1**: This week, we will cover how iOS software is developed, QA'd, and deployed. By the end of this

>   session, all participants should be able to pull the latest code from [Eigen](https://github.com/artsy/eigen) and

Reviewer: non-Artsy readers might appreciate some context Eigen and Emission. What are these two projects and why do they need to be linked?

  [Emission](https://github.com/artsy/emission), see their work in an iOS simulator, and link the two projects
  together.
- **Week 2**: This week, we will cover what makes React Native distinct from React on the web, as well as how Artsy

>   leverages shared infrastructure (such as [Palette](https://github.com/artsy/palette)) to make it easier for

Reviewer: ```suggestion
  leverages shared infrastructure (such as our design-system [Palette](https://github.com/artsy/palette)) to make it easier for
```

Similar feedback to the previous - non-Artsy readers would benefit from additional context about our projects.

  engineers to work in either one.
- **Week 3**: This week, we will cover how to create a new view controller. View controllers are the main unit of
  composition for native iOS apps, and we integrate our "Scene" React components _as_ view controllers. This
  includes routing between view controller, from both native Objective-C and React Native code.
- **Week 4**: This week, we will create our own React component to fit within the new view controller from Week 3.

>   This will be a Relay container, fetching data from [Metaphysics](https://github.com/artsy/metaphysics) v2. We

Reviewer: ```suggestion
  This will be a Relay container, fetching data from our main API, [Metaphysics](https://github.com/artsy/metaphysics). We
```

All of the learning materials
[are open source](https://github.com/artsy/README/tree/master/resources/mobile/learning-group). While the materials
are mostly specific to Artsy, they may be of help to others. And regardless, we want to adhere to our
[Open Source by Default](https://github.com/artsy/README/blob/master/culture/engineering-principles.md#open-source-by-default)
principle.

We learned quite a lot from delivering this curriculum – lessons we can apply to our next learning group:

- Learners appreciated the weekly schedule set upfront, affirming what we learned from previous learning groups.
- Learners appreciated having access to the materials ahead of time.
- Learners appreciated having the sessions recorded, to be reviewed later (or watched, in case they missed the
  session).
- Learners appreciated having office hours available; even though the office hours weren't well-attended, learners
  appreciated having access to them if they needed to.
- Learners even appreciated the awkward silences while I waited for an answer to my questions. (One survey
  respondent described it as "like pulling teeth, but helpful.")

> - Learners are varied in how they want to learn. Some liked going through things together. Some thought we went too

Reviewer: I'm in the process of reading "Make It Stick" right now, and I remember a chapter that talked about different styles of learning. My takeaway from the chapter was to push people toward the learning styles recommended by the book, as "learning styles" often means "what I am comfortable with, regardless of how effective it may be". I'm curious (a) if you remember that chapter, (b) you had different takeaways, and (c) if it affects how you think about this last bullet point.

Author: Hmm! I do remember, and I agreed with it (from my own learning experiences). I think there's room on this last bullet point to differentiate learning _styles_ (auditory, visual, etc) from the activities we do to learn (review material beforehand? etc). I think there's room on the latter, to support both people who want to come prepared and those who don't have the time. I'll revisit 👍

Author: This is worth calling out, but I don't want to distract either. I've added a paragraph below that is sufficiently hand-wave-y to acknowledge this, without diving too deeply. Thoughts?

Reviewer: I'm sorry if you lost cycles on this - I thought I edited my comment to include a statement that I wasn't expecting any changes to the article based on my comment. What you added is great, though!

  slow. Still others thought we should expect learners to do more work ahead of class.

As I enter a new chapter of my own career, leading Artsy's new Mobile Experience team, it was helpful to return to
some fundamentals; to get familiar with technology choices that we made years ago

> ([choices which I originall resisted](https://ashfurrow.com/blog/swift-vs-react-native-feels/)); and to learn from

Reviewer 2: ```suggestion
([choices which I originally resisted](https://ashfurrow.com/blog/swift-vs-react-native-feels/)); and to learn from
```

---
layout: epic
title: "iOS Learning Group"
date: "2019-12-03"
author: [ash]
categories: [ios, learning, culture]
---

Regular readers of our blog might be familiar with [Culture Amp](https://www.cultureamp.com), a tool Artsy uses to
collect anonymous feedback and take action on cultural issues (we most recently discussed the tool
[in this blog post](https://artsy.github.io/blog/2019/04/19/having-a-coffee-with-every-engineer/)). At a
company-wide level, Culture Amp has helped guide everything from Artsy's evolving culture, to our physical work
spaces, to our support for remote work. At an engineering-team level, we've also been using Culture Amp to guide
our choices in technology, documentation, and training.

In this blog, post, I'll be detailing a recent learning course we ran to share knowledge about how Artsy builds iOS
software for our entire engineering team.

<!-- more -->

Let's start at the beginning. Earlier this year, Artsy Engineering ran a survey through Culture Amp to get answers
to the following questions:

1. What is our team's opinion on our current technology choices?
2. What is our team's familiarity with or preparedness for our current technology?

> 3. Where are the areas of strength/opportunities for learning and teaching?

Author: I agree! I copied these from the original brief in Notion, so I'm on the fence about changing them for clarify. I'll take a look at rewording.

I set out to create a learning plan. I wasn't starting out from scratch – we have already run a few learning
groups, on topics ranging from Scala fundamentals to React Hooks. We used the lessons learned from _those_
experiences to define and deliver a learning plan.

I started by booking five sessions, spread out by a week. I picked a time that was a good fit for as many engineers
as possible, and I scheduled them a month ahead to give people a chance to move their schedules around. Scheduling
them up front was important, to give learners a sense of what to expect; previous learning groups had suffered from
inconsistent schedules, which led to intermittent attendance. I also made sure to go to each Tech Lead and ask them
to make sure at least one engineer from their team attended.

Next was the actual curriculum. In the spirit of "proudly discovered elsewhere", we actually looked at using
existing learning resources that someone else had already developed. However, since
[our iOS stack is a bit unique](https://artsy.github.io/series/react-native-at-artsy/) and I already have

> experience developing learning curricula, we decided to make our own. With the help of our Peer Learning Working

Reviewer 3: Might wanna leave the 'and I already have experience developing learning curricula,' clause out since it's kinda beside the point and the word 'curricula' is distracting IMO

helpful while teaching engineers at Artsy about how we build iOS software:

- If someone tries to do something themselves _before_ being told how to do it, the attempt will strengthen their
  understanding of the underlying concept. To put this into practice, I would often ask learners questions that I
  didn't expect they could answer yet, and the resulting discussion was always worthwhile.
- Interleaving different concepts together helps learners form connections between those concepts. This was
  especially important, since a big motivator for using React Native at Artsy was to share skills between web and
  iOS codebases. As an example of putting this into practice, I interleaved a discussion of
  [Relay](https://relay.dev) into our curriculum; I hoped to show learners both a new perspective of Relay, as well
  as show them how familiar writing React Native code was to writing React web code.
- Allowing for some forgetting to take place before reviewing concepts will help strengthen learner's
  understanding. To put this into practice, I would return to topics from a few weeks ago to cement their

>   understanding with learners.

Reviewer 3: might be helpful to link to an article on spaced repetition here.
